File: sdk/villow-0.1.0/villow-0.1.0/src/villow/context.py
from __future__ import annotations

from typing import Any
from uuid import uuid4

from villow.tools import Tools


class Context:
    def __init__(
        self,
        *,
        agent,
        task_id: str,
        task_template: str,
        inputs: dict[str, Any],
        answers: dict[str, Any] | None = None,
        delegated_decisions: dict[str, bool] | None = None,
        tool_access_grants: list[dict[str, Any] | str] | None = None,
        callback_urls: dict[str, str] | None = None,
        sample_mode: bool = False,
        budget_cap: dict[str, Any] | None = None,
        session_context: dict[str, Any] | None = None,
        agent_state: dict[str, Any] | None = None,
    ) -> None:
        self.agent = agent
        self.task_id = task_id
        self.task_template = task_template
        self.inputs = inputs
        self.answers = answers or {}
        self.delegated_decisions = delegated_decisions or {}
        self.tool_access_grants = tool_access_grants or []
        self.callback_urls = callback_urls or {}
        self.sample_mode = sample_mode
        self.budget_cap = budget_cap or {}
        # M6: platform-replayed session memory for follow-up units. Empty for a fresh task.
        # The publisher stays stateless — read prior transcript/artifacts from here.
        self.session_context = session_context or {}
        # N1.5: the platform replays the opaque continuation blob this agent returned on a prior
        # unit (empty for a fresh session). The platform stores/replays it but never interprets it;
        # call set_agent_state(...) to update it for the next unit.
        self.agent_state = agent_state or {}
        self._next_agent_state: dict[str, Any] | None = None
        self.emitted_events: list[dict[str, Any]] = []
        self.tools = Tools(self)

    def set_agent_state(self, state: dict[str, Any]) -> None:
        """Set the opaque continuation blob to return from this execute/result.

        The platform stores it (session-scoped, encrypted, size-capped) and replays it as
        ``ctx.agent_state`` on the next unit. Must be a JSON object; keep it small.
        """

        if not isinstance(state, dict):
            raise ValueError("agent_state must be a JSON object (dict)")
        self._next_agent_state = state

    def has_answer(self, question_id: str) -> bool:
        return question_id in self.answers

    def answer(self, question_id: str, default: Any = None) -> Any:
        return self.answers.get(question_id, default)

    def was_delegated(self, question_id: str) -> bool:
        return bool(self.delegated_decisions.get(question_id))

    def ready_to_authorize(
        self,
        *,
        normalized_inputs: dict[str, Any],
        composition_response: dict[str, Any] | None = None,
        price_units: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        payload: dict[str, Any] = {
            "task_id": self.task_id,
            "state": "ready_to_authorize",
            "normalized_inputs": normalized_inputs,
            "price_units": price_units or {},
        }
        if composition_response is not None:
            payload["composition_response"] = composition_response
        return payload

    def request_clarification(
        self,
        questions: list[dict[str, Any]],
        *,
        composition_response: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        _validate_questions(questions)
        payload: dict[str, Any] = {
            "task_id": self.task_id,
            "state": "clarification_required",
            "phase": "pre_execution",
            "clarification_id": f"clarification_{uuid4().hex}",
            "questions": questions,
        }
        if composition_response is not None:
            payload["composition_response"] = composition_response
        return payload

    async def request_mid_execution_clarification(self, questions: list[dict[str, Any]]) -> dict[str, Any]:
        _validate_questions(questions)
        payload = {
            "task_id": self.task_id,
            "state": "clarification_required",
            "phase": "mid_execution",
            "clarification_id": f"clarification_{uuid4().hex}",
            "questions": questions,
        }
        return await self._emit_callback("clarification_required", payload)

    async def report_progress(self, event: dict[str, Any] | None = None, **progress: Any) -> dict[str, Any]:
        payload = {
            "task_id": self.task_id,
            "state": "executing",
            "progress": event or progress,
        }
        return await self._emit_callback("status", payload)

    async def report_sample_complete(self, sample_artifacts: list[dict[str, Any]]) -> dict[str, Any]:
        payload = {"task_id": self.task_id, "state": "sample_review", "sample_artifacts": sample_artifacts}
        return await self._emit_callback("sample_complete", payload)

    async def stage_artifact(self, artifact: dict[str, Any]) -> dict[str, Any]:
        payload = {"task_id": self.task_id, "state": "artifact_review", "artifact": artifact}
        return await self._emit_callback("artifact_staged", payload)

    async def report_unit_charge(self, unit_charge: dict[str, Any]) -> dict[str, Any]:
        payload = {"task_id": self.task_id, "state": "executing", "cost_telemetry": unit_charge}
        return await self._emit_callback("status", payload)

    async def report_cost_telemetry(self, cost_telemetry: dict[str, Any]) -> dict[str, Any]:
        """Deprecated compatibility alias for report_unit_charge."""

        return await self.report_unit_charge(cost_telemetry)

    def resolve_tool_access_grant(self, tool_name: str) -> str | None:
        if not self.tool_access_grants:
            return None
        prefix = f"{tool_name.split('.', 1)[0]}." if "." in tool_name else tool_name
        for grant in self.tool_access_grants:
            if isinstance(grant, str):
                return grant
            grant_tool = str(grant.get("tool_name") or "")
            grant_id = grant.get("tool_access_grant_id") or grant.get("grant_id") or grant.get("id")
            if grant_tool == tool_name and grant_id:
                return str(grant_id)
        for grant in self.tool_access_grants:
            if isinstance(grant, dict):
                grant_tool = str(grant.get("tool_name") or "")
                grant_id = grant.get("tool_access_grant_id") or grant.get("grant_id") or grant.get("id")
                if grant_tool.startswith(prefix) and grant_id:
                    return str(grant_id)
        legacy_ids: list[str] = []
        for grant in self.tool_access_grants:
            if isinstance(grant, dict):
                grant_id = grant.get("tool_access_grant_id") or grant.get("grant_id") or grant.get("id")
                if grant_id and not grant.get("tool_name"):
                    legacy_ids.append(str(grant_id))
        if len(legacy_ids) == 1:
            return legacy_ids[0]
        return None

    async def _emit_callback(self, kind: str, payload: dict[str, Any]) -> dict[str, Any]:
        self.emitted_events.append({"kind": kind, "payload": payload})
        url = self.callback_urls.get(kind)
        if not url:
            return {"accepted": True, "local": True}
        return await self.agent.signed_platform_post(url, payload, idempotency_key=self._idempotency_key(kind))

    def _idempotency_key(self, operation: str) -> str:
        return f"{self.task_id}:{operation}:{uuid4().hex}"


def _validate_questions(questions: list[dict[str, Any]]) -> None:
    for question in questions:
        if "decide_for_me" not in question or question["decide_for_me"] is None:
            raise ValueError("every clarification question requires decide_for_me")

File: sdk/villow-sdk-0.1.0/package/src/context.ts
import { randomUUID } from 'node:crypto';

import type { Agent } from './agent.js';
import { Tools } from './tools.js';

export interface ContextOptions {
  agent: Agent;
  taskId: string;
  taskTemplate: string;
  inputs: Record<string, unknown>;
  answers?: Record<string, unknown>;
  delegatedDecisions?: Record<string, boolean>;
  toolAccessGrants?: Array<Record<string, unknown> | string>;
  callbackUrls?: Record<string, string>;
  sampleMode?: boolean;
  budgetCap?: Record<string, unknown>;
  sessionContext?: Record<string, unknown>;
  agentState?: Record<string, unknown>;
}

export class Context {
  agent: Agent;
  taskId: string;
  taskTemplate: string;
  inputs: Record<string, unknown>;
  answers: Record<string, unknown>;
  delegatedDecisions: Record<string, boolean>;
  toolAccessGrants: Array<Record<string, unknown> | string>;
  callbackUrls: Record<string, string>;
  sampleMode: boolean;
  budgetCap: Record<string, unknown>;
  // M6: platform-replayed session memory for follow-up units (empty for a fresh task).
  // The publisher stays stateless — read prior transcript/artifacts from here.
  sessionContext: Record<string, unknown>;
  // N1.5: the platform replays the opaque continuation blob this agent returned on a prior unit
  // (empty for a fresh session). The platform stores/replays it but never interprets it; call
  // setAgentState(...) to update it for the next unit.
  agentState: Record<string, unknown>;
  nextAgentState: Record<string, unknown> | null = null;
  emittedEvents: Array<{ kind: string; payload: Record<string, unknown> }> = [];
  tools: Tools;

  constructor(options: ContextOptions) {
    this.agent = options.agent;
    this.taskId = options.taskId;
    this.taskTemplate = options.taskTemplate;
    this.inputs = options.inputs;
    this.answers = options.answers ?? {};
    this.delegatedDecisions = options.delegatedDecisions ?? {};
    this.toolAccessGrants = options.toolAccessGrants ?? [];
    this.callbackUrls = options.callbackUrls ?? {};
    this.sampleMode = options.sampleMode ?? false;
    this.budgetCap = options.budgetCap ?? {};
    this.sessionContext = options.sessionContext ?? {};
    this.agentState = options.agentState ?? {};
    this.tools = new Tools(this);
  }

  /**
   * Set the opaque continuation blob to return from this execute/result. The platform stores it
   * (session-scoped, encrypted, size-capped) and replays it as `ctx.agentState` on the next unit.
   * Must be a JSON object; keep it small.
   */
  setAgentState(state: Record<string, unknown>): void {
    if (!state || typeof state !== 'object' || Array.isArray(state)) {
      throw new Error('agentState must be a JSON object');
    }
    this.nextAgentState = state;
  }

  hasAnswer(questionId: string): boolean {
    return Object.hasOwn(this.answers, questionId);
  }

  answer(questionId: string, defaultValue?: unknown): unknown {
    return this.answers[questionId] ?? defaultValue;
  }

  wasDelegated(questionId: string): boolean {
    return Boolean(this.delegatedDecisions[questionId]);
  }

  readyToAuthorize(input: {
    normalizedInputs: Record<string, unknown>;
    compositionResponse?: Record<string, unknown>;
    priceUnits?: Record<string, unknown>;
  }): Record<string, unknown> {
    const payload: Record<string, unknown> = {
      task_id: this.taskId,
      state: 'ready_to_authorize',
      normalized_inputs: input.normalizedInputs,
      price_units: input.priceUnits ?? {},
    };
    if (input.compositionResponse) payload.composition_response = input.compositionResponse;
    return payload;
  }

  requestClarification(questions: Array<Record<string, unknown>>, options: { compositionResponse?: Record<string, unknown> } = {}): Record<string, unknown> {
    validateQuestions(questions);
    const payload: Record<string, unknown> = {
      task_id: this.taskId,
      state: 'clarification_required',
      phase: 'pre_execution',
      clarification_id: `clarification_${randomUUID().replace(/-/g, '')}`,
      questions,
    };
    if (options.compositionResponse) payload.composition_response = options.compositionResponse;
    return payload;
  }

  async requestMidExecutionClarification(questions: Array<Record<string, unknown>>): Promise<Record<string, unknown>> {
    validateQuestions(questions);
    return this.emitCallback('clarification_required', {
      task_id: this.taskId,
      state: 'clarification_required',
      phase: 'mid_execution',
      clarification_id: `clarification_${randomUUID().replace(/-/g, '')}`,
      questions,
    });
  }

  async reportProgress(event: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.emitCallback('status', { task_id: this.taskId, state: 'executing', progress: event });
  }

  async reportSampleComplete(sampleArtifacts: Array<Record<string, unknown>>): Promise<Record<string, unknown>> {
    return this.emitCallback('sample_complete', { task_id: this.taskId, state: 'sample_review', sample_artifacts: sampleArtifacts });
  }

  async stageArtifact(input: Record<string, unknown>): Promise<Record<string, unknown>> {
    const artifact = Object.hasOwn(input, 'artifact') ? input.artifact : input;
    return this.emitCallback('artifact_staged', { task_id: this.taskId, state: 'artifact_review', artifact });
  }

  async reportUnitCharge(unitCharge: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.emitCallback('status', { task_id: this.taskId, state: 'executing', cost_telemetry: unitCharge });
  }

  /** @deprecated Use reportUnitCharge. */
  async reportCostTelemetry(costTelemetry: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.reportUnitCharge(costTelemetry);
  }

  resolveToolAccessGrant(toolName: string): string | null {
    if (!this.toolAccessGrants.length) return null;
    const prefix = toolName.includes('.') ? `${toolName.split('.', 1)[0]}.` : toolName;
    for (const grant of this.toolAccessGrants) {
      if (typeof grant === 'string') return grant;
      const grantTool = String(grant.tool_name ?? '');
      const grantId = grant.tool_access_grant_id ?? grant.grant_id ?? grant.id;
      if (grantTool === toolName && grantId) return String(grantId);
    }
    for (const grant of this.toolAccessGrants) {
      if (typeof grant !== 'object' || grant === null) continue;
      const grantTool = String(grant.tool_name ?? '');
      const grantId = grant.tool_access_grant_id ?? grant.grant_id ?? grant.id;
      if (grantTool.startsWith(prefix) && grantId) return String(grantId);
    }
    const legacyIds: string[] = [];
    for (const grant of this.toolAccessGrants) {
      if (typeof grant !== 'object' || grant === null) continue;
      const grantId = grant.tool_access_grant_id ?? grant.grant_id ?? grant.id;
      if (grantId && !grant.tool_name) legacyIds.push(String(grantId));
    }
    if (legacyIds.length === 1) return legacyIds[0]!;
    return null;
  }

  idempotencyKey(operation: string): string {
    return `${this.taskId}:${operation}:${randomUUID().replace(/-/g, '')}`;
  }

  private async emitCallback(kind: string, payload: Record<string, unknown>): Promise<Record<string, unknown>> {
    this.emittedEvents.push({ kind, payload });
    const url = this.callbackUrls[kind];
    if (!url) return { accepted: true, local: true };
    return this.agent.signedPlatformPost(url, payload, { idempotencyKey: this.idempotencyKey(kind) });
  }
}

function validateQuestions(questions: Array<Record<string, unknown>>): void {
  for (const question of questions) {
    if (!Object.hasOwn(question, 'decide_for_me') || question.decide_for_me === undefined || question.decide_for_me === null) {
      throw new Error('every clarification question requires decideForMe');
    }
  }
}
